tooling/templatize/pkg/config/types.go

```diff
@@ -0,0 +1,81 @@
+package config
+
+import "fmt"
+
+type configProviderImpl struct {
+	config string
+}
+
+type Variables map[string]interface{}
+
+func (v Variables) AddNested(key string, other map[string]interface{}) error {
+	if _, exists := v[key]; exists {
+		return fmt.Errorf("%s exists already in Variables", key)
+	}
+	other[key] = other
+	return nil
+}
+
+type VariableOverrides struct {
```

Comment:
Structurally, what is the difference between `VariableOverrides`, `CloudVariableOverride` and `DeployEnvVariableOverride`? Why do we need three types?

Reply:
Why do the `*Override` structs have exported members if it seems like you need to use the getters for correctly traversing these? Perhaps an unexported struct you could unmarshal into + an interface would be better if you want to guarantee callers use the getters.

Reply:
> Why do we need three types?

their structure is the same but the struct tags for yaml parsing are different. i made the cloud and deployenv ones directly nested types within the outer one. let me know what you think.

> exported members

i made everything non-exported and defined accessors on an interface